src/GafferArnold/ArnoldImager.cpp

```diff
@@ -0,0 +1,227 @@
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "GafferArnold/ArnoldImager.h"
+
+#include "GafferScene/Shader.h"
+#include "GafferScene/ShaderPlug.h"
+
+#include "Gaffer/StringPlug.h"
+
+#include "IECoreScene/ShaderNetwork.h"
+#include "IECoreScene/ShaderNetworkAlgo.h"
+
+using namespace IECore;
+using namespace IECoreScene;
+using namespace Gaffer;
+using namespace GafferScene;
+using namespace GafferArnold;
+
+namespace
+{
+
+const InternedString g_inputParameterName( "input" );
+const InternedString g_imagerAttributeName( "ai:imager" );
+const InternedString g_imagerOptionName( "option:ai:imager" );
+
+ShaderNetwork::Parameter firstInput( const ShaderNetwork *network, const InternedString &shader )
+{
+	ShaderNetwork::Parameter result( shader, g_inputParameterName );
+	while( true )
+	{
+		if( auto input = network->input( result ) )
+		{
+			result.shader = input.shader;
+		}
+		else
+		{
+			return result;
+		}
+	}
+}
+
+} // namespace
+
+GAFFER_NODE_DEFINE_TYPE( ArnoldImager );
+
+size_t ArnoldImager::g_firstPlugIndex = 0;
+
+ArnoldImager::ArnoldImager( const std::string &name )
+	:	GlobalsProcessor( name )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+	addChild( new ShaderPlug( "imager" ) );
+	addChild( new IntPlug( "mode", Plug::In, (int)Mode::Replace, (int)Mode::Replace, (int)Mode::InsertLast ) );
+}
+
+ArnoldImager::~ArnoldImager()
+{
+}
+
+GafferScene::ShaderPlug *ArnoldImager::imagerPlug()
+{
+	return getChild<ShaderPlug>( g_firstPlugIndex );
+}
+
+const GafferScene::ShaderPlug *ArnoldImager::imagerPlug() const
+{
+	return getChild<ShaderPlug>( g_firstPlugIndex );
+}
+
+Gaffer::IntPlug *ArnoldImager::modePlug()
+{
+	return getChild<IntPlug>( g_firstPlugIndex + 1 );
+}
+
+const Gaffer::IntPlug *ArnoldImager::modePlug() const
+{
+	return getChild<IntPlug>( g_firstPlugIndex + 1 );
+}
+
+bool ArnoldImager::acceptsInput( const Gaffer::Plug *plug, const Gaffer::Plug *inputPlug ) const
+{
+	if( !GlobalsProcessor::acceptsInput( plug, inputPlug ) )
+	{
+		return false;
+	}
+
+	if( plug != imagerPlug() )
+	{
+		return true;
+	}
+
+	if( !inputPlug )
+	{
+		return true;
+	}
+
+	const Plug *sourcePlug = inputPlug->source();
+	auto *sourceShader = runTimeCast<const GafferScene::Shader>( sourcePlug->node() );
+	if( !sourceShader )
+	{
+		return true;
```

Comment:
I'm trying to follow this logic - under what circumstances would this happen?  Why is it correct to accept a connection from something other than a Shader?

Reply:
For expressions, promotion etc...